anycloud/node/src/datastore.ts

```diff
@@ -30,65 +34,94 @@ export class DataStore {
       if (dsKey in this.localDS) {
         const val = this.localDS[dsKey];
-        return JSON.stringify(val);
+        return JSON.parse(val) || val;
       } else {
         return undefined;
       }
     }
     try {
-      return JSON.stringify(await this.request('GET', `${this.ctrlPortUrl}get/${dsKey.toString()}`));
+      const dsValue = await this.request('GET', `${this.ctrlPortUrl}get/${dsKey}`);
+      return JSON.parse(dsValue) || dsValue;
     } catch (e) {
       return e;
     }
   }
 
   async set(dsKey: string, dsValue: any): Promise<string> {
-    if (!dsKey) return 'fail';
-    const parsedDSValue = (function () {
+    if (DataStore.isInvalidKey(dsKey)) return 'fail';
+    const maybeStringifiedValue = (function () {
       try {
         return JSON.stringify(dsValue);
       } catch (_) {
-        return null;
+        return dsValue;
       }
     })();
-    if (parsedDSValue === null) return 'fail';
     if (this.isLocal) {
-      this.localDS[dsKey] = parsedDSValue;
+      this.localDS[dsKey] = maybeStringifiedValue;
       return 'ok';
     }
     try {
-      return await this.request('POST', `${this.ctrlPortUrl}set/${dsKey.toString()}`, parsedDSValue);
+      return await this.request('POST', `${this.ctrlPortUrl}set/${dsKey}`, maybeStringifiedValue);
     } catch (_) {
       return 'fail';
     }
   }
 
   async del(dsKey: string): Promise<boolean> {
-    if (!dsKey) return false;
+    if (DataStore.isInvalidKey(dsKey)) return false;
     if (this.isLocal) {
       return dsKey in this.localDS ? delete this.localDS[dsKey] : false;
     }
     try {
-      return await this.request('GET', `${this.ctrlPortUrl}del/${dsKey.toString()}`) === 'true';
+      return await this.request('GET', `${this.ctrlPortUrl}del/${dsKey}`) === 'true';
     } catch (_) {
       return false;
     }
   }
 
   async has(dsKey: string): Promise<boolean> {
-    if (!dsKey) return false;
+    if (DataStore.isInvalidKey(dsKey)) return false;
     if (this.isLocal) {
       return dsKey in this.localDS;
     }
     try {
-      return await this.request('GET', `${this.ctrlPortUrl}has/${dsKey.toString()}`) === 'true';
+      return await this.request('GET', `${this.ctrlPortUrl}has/${dsKey}`) === 'true';
     } catch (_) {
       return false;
     }
   }
 }
+
+const dsHandler = () => {
+  const ds = new DataStore();
+  return {
+    get: async (_, dsKey: string): Promise<string> => {
+      if (DataStore.isInvalidKey(dsKey)) {
+        return undefined;
+      }
+      return await ds.get(dsKey);
+    },
+    set: (_, dsKey: string, dsValue: any): boolean => {
+      if (DataStore.isInvalidKey(dsKey)) {
+        return false;
+      }
+      ds.set(dsKey, dsValue);
+      return true;
+    },
+    deleteProperty: (_, dsKey: string): boolean => {
+      if (DataStore.isInvalidKey(dsKey)) {
+        return false;
+      }
+      ds.del(dsKey);
+      return true;
+    }
+  };
+};
+
+export const datastore = new Proxy({}, dsHandler());
```

Comment:
I wouldn't mind also exporting an instance of the `DataStore` class (Not sure about that name, though. It's technically 2 words in English, but "datastore" is one logical name in our API, I'd just call it the `Datastore` class.)

Maybe the proxy version gets the name `ds` and the class instance `datastore`?

Reply:
I have struggling with the naming haha, because we define the name here but later we have to give another one for the export in the `index.ts`. 
So what do you think of this:
`datastore.ts`:
- proxy: `dsProxy`
- class: `DatastoreClass`

`index.ts`:
```
export const DataStore = DatastoreClass;
export const ds = dsProxy;
```

Reply:
So that second export in `index.ts` doesn't need to be a rename at all. And we don't have to export the raw class, just the instance of the class. I'd just do:

`datastore.ts`
```ts
export const datastore = new Datastore();
export const ds = new Proxy({}, dsHandler());
```

`index.ts`
```ts
import { datastore, ds } from './datastore'
export datastore
export ds
```

Reply:
https://stackoverflow.com/a/30714301